src/compiler/tsbuildPublic.ts

```diff
@@ -1494,80 +1541,199 @@ namespace ts {
-            if (extendedConfigStatus) return extendedConfigStatus;
-
-            // Check package file time
-            const dependentPackageFileStatus = forEach(
-                state.lastCachedPackageJsonLookups.get(resolvedPath) || emptyArray,
-                ([path]) => checkConfigFileUpToDateStatus(state, path, oldestOutputFileTime, oldestOutputFileName)
-            );
-            if (dependentPackageFileStatus) return dependentPackageFileStatus;
+
+        // Now see if all outputs are newer than the newest input
+        // Dont check output timestamps if we have buildinfo telling us output is uptodate
+        if (!buildInfoPath) {
+            // Collect the expected outputs of this project
+            const outputs = getAllProjectOutputs(project, !host.useCaseSensitiveFileNames());
+            const outputTimeStampMap = getOutputTimeStampMap(state, resolvedPath);
+            for (const output of outputs) {
+                const path = toPath(state, output);
+                // Output is missing; can stop checking
+                let outputTime = outputTimeStampMap?.get(path);
+                if (!outputTime) {
+                    outputTime = ts.getModifiedTime(state.host, output);
+                    outputTimeStampMap?.set(path, outputTime);
+                }
+
+                if (outputTime === missingFileModifiedTime) {
+                    return {
+                        type: UpToDateStatusType.OutputMissing,
+                        missingOutputFileName: output
+                    };
+                }
+
+                if (outputTime < oldestOutputFileTime) {
+                    oldestOutputFileTime = outputTime;
+                    oldestOutputFileName = output;
+                }
+
+                // If an output is older than the newest input, we can stop checking
+                if (outputTime < newestInputFileTime) {
+                    return {
+                        type: UpToDateStatusType.OutOfDateWithSelf,
+                        outOfDateOutputFileName: oldestOutputFileName,
+                        newerInputFileName: newestInputFileName
+                    };
+                }
+            }
         }
 
-        if (!force && !state.buildInfoChecked.has(resolvedPath)) {
-            state.buildInfoChecked.set(resolvedPath, true);
-            const buildInfoPath = getTsBuildInfoEmitOutputFilePath(project.options);
-            if (buildInfoPath) {
-                const value = state.readFileWithCache(buildInfoPath);
-                const buildInfo = value && getBuildInfo(value);
-                if (buildInfo && (buildInfo.bundle || buildInfo.program) && buildInfo.version !== version) {
+        const seenRefs = buildInfoPath ? new Set<ResolvedConfigFilePath>() : undefined;
+        const buildInfoCacheEntry = state.buildInfoCache.get(resolvedPath);
+        seenRefs?.add(resolvedPath);
+
+        let pseudoUpToDate = false;

```

Comment:
As discussed offline, it would be helpful to have a description somewhere in the code of what "pseudo up-to-date" means.